src-plugins/mscAlgorithmPaint/mscAlgorithmPaintToolBox.cpp

```diff
@@ -1445,44 +1451,26 @@ void AlgorithmPaintToolBox::addSliceToStack(medAbstractView * view,const unsigne
             m_redoStacks->value(view)->clear();
 }
 
-void AlgorithmPaintToolBox::closeView()
+void AlgorithmPaintToolBox::clear()
 {
-    // close every data in the view
-    medAbstractView* view = qobject_cast<medAbstractView*>(QObject::sender());
-    medAbstractLayeredView* layeredView = dynamic_cast<medAbstractLayeredView*>(view);
-
-    if (layeredView->layersCount() > 0)
+    if (currentView && (currentView->layersCount()>0))
     {
-        // remove a data, which will recursively call this closeView() function
-        // until there is no more data in the view
-        layeredView->removeLayer(0);
-    }
-    else
-    {
-        // clean the toolbox when there is no data left in the view
-        onViewClosed(view);
-    }
-}
-
-void AlgorithmPaintToolBox::onViewClosed(medAbstractView* viewClosed)
-{
-    clearMask();
-
-    if (m_undoStacks->value(viewClosed))
-    {
-        m_undoStacks->value(viewClosed)->clear();
-        m_redoStacks->value(viewClosed)->clear();
-        m_undoStacks->remove(viewClosed);
-        m_redoStacks->remove(viewClosed);
+        // remove the "no name" mask data in the view
+        for(int unsigned i=0; i<currentView->layersCount(); i++)
+        {
+            if (currentView->layerData(i)->identifier() == "medImageMaskAnnotationData")
+            {
+                currentView->removeLayer(i); // call clear() again
+                return;
```

Comment:
I hate to be annoying here but ```clear``` should clear the toolbox in all cases, not do something different depending on the state of the toolbox. Just because it happens to be called twice when you click on clear is irrelevant. If someone changes that someday then this will not work anymore. I think you should replace the return with a break so that ```clearMask``` and ```updateView``` is always called, and we don't care if it get's called twice since it won't cause any noticeable delay.